docs/developer-guide/performance.md

```diff
@@ -30,402 +30,835 @@ Some good places to check for performance improvements are:
+    layers={layers}
+  />;
+}
+```
 
-  function render() {
-    const layer = new ScatterplotLayer({
-      data: filteredData,
-      ...
-    });
+  </TabItem>
+</Tabs>
 
-    deck.setProps({layers: [layer]});
-  }
-  ```
 
 #### Use updateTriggers
 
-  So `data` has indeed changed. Do we have an entirely new collection of objects? Or did just certain fields changed in each row? Remember that changing `data` will update *all* buffers, so if, for example, object positions have not changed, it will be a waste of time to recalculate them.
+So `data` has indeed changed. Do we have an entirely new collection of objects? Or did just certain fields changed in each row? Remember that changing `data` will update *all* buffers, so if, for example, object positions have not changed, it will be a waste of time to recalculate them.
 
-  ```js
-  // Bad
-  const DATA = [...];
-  let currentYear = null;
-  let currentData = DATA;
+<Tabs groupId="language">
+  <TabItem value="ts" label="TypeScript">
 
-  function selectYear(year) {
-    currentYear = year;
-    currentData = DATA.map(d => ({
-      position: d.position,
-      population: d.populationsByYear[year]
-    }));
-    render();
-  }
+```ts
+import {Deck} from '@deck.gl/core';
+import {ScatterplotLayer} from '@deck.gl/layers';
 
-  function render() {
-    const layer = new ScatterplotLayer({
-      // `data` changes every time year changed, but positions don't need to update
-      data: currentData,
-      getPosition: d => d.position,
-      getRadius: d => Math.sqrt(d.population),
-      ...
-    });
-
-    deck.setProps({layers: [layer]});
-  }
-  ```
+type CensusTract = {
+  centroid: [longitude: number, latitude: number];
+  populationsByYear: {[year: number]: number};
+};
+const DATA: CensusTract[] = [...];
 
-  In this case, it is more efficient to use [`updateTriggers`](../api-reference/core/layer.md#updatetriggers) to invalidate only the selected attributes:
+const deckInstance = new Deck({...});
+```
 
-  ```js
-  // Good
-  const DATA = [...];
-  let currentYear = null;
+```ts title="Bad practice"
+type CensusTractCurrentYear = {
+  centroid: [longitude: number, latitude: number];
+  population: number;
+};
 
-  function selectYear(year) {
-    currentYear = year;
-    render();
-  }
-
-  function render() {
-    const layer = new ScatterplotLayer({
+function render(year: number) {
+  const layers = [
+    new ScatterplotLayer<CensusTractCurrentYear>({
+      // `data` changes every time year changed, but positions don't need to update
+      data: DATA.map(d => ({
+        centroid: d.centroid,
+        population: d.populationsByYear[year]
+      })),
+      getPosition: (d: CensusTractCurrentYear) => d.centroid,
+      getRadius: (d: CensusTractCurrentYear) => Math.sqrt(d.population)
+    })
+  ];
+
+  deckInstance.setProps({layers});
+}
+```
+
+In this case, it is more efficient to use `updateTriggers` to invalidate only the selected attributes:
+
+```ts title="Good practice"
```

Comment:
What does title do? Should it be on the react one as well?

Reply:
It shows up as a header before the code block. I use the headers "bad practice"/"good practice" here because not all code samples on this page are for copy and paste.